# HG changeset patch
# User Rik <rik@octave.org>
# Date 1534540368 25200
#      Fri Aug 17 14:12:48 2018 -0700
# Node ID 94b8dc8292af6739de898f5d2a05d7ab20e92c4a
# Parent  650e5d50cd4787917cca0d7779772d1c0352a91a
# Parent  7233dae64579ebf1ad5f3d66d0fd22157f01b3d8
maint: merge stable to default.

diff --git a/libgui/src/m-editor/file-editor-tab.cc b/libgui/src/m-editor/file-editor-tab.cc
--- a/libgui/src/m-editor/file-editor-tab.cc
+++ b/libgui/src/m-editor/file-editor-tab.cc
@@ -110,19 +110,16 @@ namespace octave
 
     _bp_lines.clear ();      // start with empty lists of breakpoints
     _bp_conditions.clear ();
     m_bp_restore_count = 0;
 
     // Initialize last modification date to now
     m_last_modified = QDateTime::currentDateTimeUtc();
 
-    // disable editor drag & drop so parent can handle
-    _edit_area->setAcceptDrops (false);
-
     connect (_edit_area, SIGNAL (cursorPositionChanged (int, int)),
              this, SLOT (handle_cursor_moved (int,int)));
 
     connect (_edit_area, SIGNAL (SCN_CHARADDED (int)),
              this, SLOT (handle_char_added (int)));
 
     connect (_edit_area, SIGNAL (SCN_DOUBLECLICK (int, int, int)),
              this, SLOT (handle_double_click (int, int, int)));
diff --git a/libgui/src/m-editor/octave-qscintilla.cc b/libgui/src/m-editor/octave-qscintilla.cc
--- a/libgui/src/m-editor/octave-qscintilla.cc
+++ b/libgui/src/m-editor/octave-qscintilla.cc
@@ -41,16 +41,17 @@ along with Octave; see the file COPYING.
 #include <Qsci/qscilexerbash.h>
 #include <Qsci/qscilexerperl.h>
 #include <Qsci/qscilexerbatch.h>
 #include <Qsci/qscilexerdiff.h>
 
 #include <Qsci/qscicommandset.h>
 
 #include <QKeySequence>
+#include <QMimeData>
 #include <QShortcut>
 #include <QToolTip>
 #include <QVBoxLayout>
 
 #include "gui-preferences.h"
 #include "resource-manager.h"
 #include "shortcut-manager.h"
 
@@ -957,11 +958,26 @@ namespace octave
       {
         if (first_word == "unwind_protect")
           first_word = '_' + first_word;
         next_line = "end" + first_word + "\n";
       }
 
     insertAt (QString (start, ' ') + next_line, linenr + 2, 0);
   }
+
+  void octave_qscintilla::dragEnterEvent (QDragEnterEvent *e)
+  {
+    // if is not dragging a url, pass to qscintilla to handle,
+    // otherwise ignore it so that it will be handled by
+    // the parent
+    if (!e->mimeData ()->hasUrls ())
+      {
+        QsciScintilla::dragEnterEvent (e);
+      }
+    else
+      {
+        e->ignore();
+      }
+  }
 }
 
 #endif
diff --git a/libgui/src/m-editor/octave-qscintilla.h b/libgui/src/m-editor/octave-qscintilla.h
--- a/libgui/src/m-editor/octave-qscintilla.h
+++ b/libgui/src/m-editor/octave-qscintilla.h
@@ -101,16 +101,18 @@ namespace octave
   protected:
 
     void focusInEvent (QFocusEvent *focusEvent);
 
     void show_replace_action_tooltip (void);
 
     void keyPressEvent (QKeyEvent *e);
 
+    void dragEnterEvent (QDragEnterEvent *e);
+
   private:
 
     void auto_close (int auto_endif, int l,
                      const QString& line, QString& first_word);
 
     QString m_word_at_cursor;
 
     QString m_selection;
diff --git a/liboctave/array/Sparse.h b/liboctave/array/Sparse.h
--- a/liboctave/array/Sparse.h
+++ b/liboctave/array/Sparse.h
@@ -470,17 +470,17 @@ public:
   Sparse<T>& insert (const Sparse<T>& a, const Array<octave_idx_type>& idx);
 
   bool issquare (void) const { return (dim1 () == dim2 ()); }
 
   OCTAVE_DEPRECATED (4.4, "use 'issquare' instead")
   bool is_square (void) const
   { return issquare (); }
 
-  bool isempty (void) const { return (rows () < 1 && cols () < 1); }
+  bool isempty (void) const { return (rows () < 1 || cols () < 1); }
 
   OCTAVE_DEPRECATED (4.4, "use 'isempty' instead")
   bool is_empty (void) const
   { return isempty (); }
 
   Sparse<T> transpose (void) const;
 
   T * data (void) { make_unique (); return rep->d; }
